### PR TITLE
AI Agents Spawning inside Buildings

### PR DESCRIPTION
**Pull request by AndyTechGuy**
## Contains
- Gooey characters spawning inside buildings generated by DynamicCities.
- Each spawnable character now contains an `AgentComponent`.
- Any building that is able to house characters now contains a `PotentialHomeComponent` in the building prefab, which lists the characters currently inside the building, and how many characters a building can provide shelter for.
- When a character spawns inside a building, it is given a `HomeComponent`, which provides a reference to this building.

## How to Test
- This PR depends on Terasology/DynamicCities#43, make sure this change is either merged or included manually or else **the game will crash**.
- Start a game with MetalRenegades enabled.
- When a city spawns, look inside any of the buildings. There should be any number from 1 to 5 gooey characters inside depending on the building.

### CODE REVIEW

File: src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    @In
+    private EntityManager entityManager;
+
+    @In
+    private PrefabManager prefabManager;
+
+    @Override
+    public void update(float delta) {
+        for (EntityRef entity : entityManager.getEntitiesWith(PotentialHomeComponent.class)) {
```
Comment (Cervator) [inline src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java]: Beware of using `update` methods that are un-throttled (literally run once every single game loop ticket), especially if they do something potentially expensive like a `entityManager.getEntitiesWith` lookup. That'll add up over time and there are usually more efficient alternatives :-)

One easy option is to throttle, that is simply add a time period during which you'll skip processing. Keep the last timestamp of when the logic ran then do a comparison vs current time and only execute if more than `x` seconds have passed. You _probably_ don't actually need to spawn agents 60 times per second :-)

File: src/main/java/org/terasology/metalrenegades/ai/component/AgentComponent.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+package org.terasology.metalrenegades.ai.component;
+
+import org.terasology.entitySystem.Component;
+
+/**
+ * A component added to prefabs containing spawnable AI agents.
+ */
+public class AgentComponent implements Component {
```
Comment (Cervator) [inline src/main/java/org/terasology/metalrenegades/ai/component/AgentComponent.java]: Maybe use a more specific term here than `Agent` which might be at high risk for colliding with actual AI architectural systems? How about `Resident` ?

File: src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    /**
+     * Spawns a random agent inside the center of a provided building entity.
+     *
+     * @param homeEntity The building entity to spawn inside.
+     * @return The new agent entity, or null if spawning is not possible.
+     */
+    private EntityRef spawnAgent(EntityRef homeEntity) {
+        Prefab agentPrefab = chooseAgentPrefab();
+        if(agentPrefab == null) { // if no prefab is available.
```
Comment (Cervator) [inline src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java]: Run Checkstyle over your code, it'll highlight trivial code style issues like here - convention is to have a space between `if` and the starting `(` :-)

**Review by Cervator — CHANGES_REQUESTED** [review]
Just a few relatively tiny comments and recommendations :-)

File: src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java (hunk 1 above)
Comment (AndyTechGuy) [reply src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java]: But what if I want to spawn my agents 60 times per second :)

I changed this to be on a throttle instead, set to 30 seconds between each cycle.

File: src/main/java/org/terasology/metalrenegades/ai/component/AgentComponent.java (hunk 2 above)
Comment (AndyTechGuy) [reply src/main/java/org/terasology/metalrenegades/ai/component/AgentComponent.java]: The agent component and references to 'agent' have been changed to "resident" as needed.

File: src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java (hunk 3 above)
Comment (AndyTechGuy) [reply src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java]: I have a checkstyle plugin on intelliJ, but I guess a few of them slipped through, whoops! Fixed these where I missed them 👍

**Pull request merged**